# Advanced Dataframes

## Exercises I

### 1-3. Install pymysql. Add env.py to .gitignore. Create function in env.py to create sql connection.

In [273]:
import pandas as pd
import numpy as np
import env

my_url = env.get_db_url(
    env.user,
    env.host,
    env.password,
    'employees')

url = f'mysql+pymysql://{env.user}:{env.password}@{env.host}/employees'


### 4. Use your function to obtain a connection to the employees database.



In [274]:
# pd.read_sql('literal sql syntax to pull query', connection_string)

# using function from env.py
my_url = env.get_db_url(
    env.user,
    env.host,
    env.password,
    'employees')
# method 2
url = f'mysql+pymysql://{env.user}:{env.password}@{env.host}/employees'

# write query and run query and my_url through pd.read_sql
query = 'select * from employees limit 5'
df = pd.read_sql(query, my_url)
df

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

### 5. Once you have successfully run a query:
 - a. Intentionally make a typo in the database url. What kind of error message do you see?

 - b. Intentionally make an error in your SQL query. What does the error message look like?

In [275]:
# a. ANSWER: Access denied
#typo_url = my_url + '5'
#df_t1 = pd.read_sql(query, typo_url)
#df_t1

In [276]:
# b. ANSWER: In one case I got an error message that there was an endeclared variable. 
# In another, there was no error message, it just changed the limit from 5 to 55
#typo_query = query + 't'
#df_t2 = pd.read_sql(typo_query, my_url)
#df_t2

### 6. Read the employees and titles tables into two separate DataFrames.



In [277]:
# for employees table:
# Write query and create DataFrame
query_1 = '''
select *
from
employees
'''
employees_df = pd.read_sql(query_1, my_url)

In [278]:
# for titles table:
# Write query and create DataFrame
query_2 = '''
select * 
from
titles
'''
titles_df = pd.read_sql(query_2, my_url)

### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?



In [279]:
# employees (300024, 6)
print(employees_df.shape)

# titles (443308, 4)
print(titles_df.shape)

# It makes sense that there would be more rows in titles than in employees as employees
# may have several different roles during their time at the company.

(300024, 6)
(443308, 4)


### 8. Display the summary statistics for each DataFrame.



In [280]:
# employees_df
print(employees_df.info())
print(employees_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB
None
              emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000


In [281]:
# titles_df
print(titles_df.info())
print(titles_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB
None
              emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000


### 9. How many unique titles are in the titles DataFrame?



In [282]:
titles_df.columns

Index(['emp_no', 'title', 'from_date', 'to_date'], dtype='object')

In [283]:
len(pd.unique(titles_df['title']))

7

### 10. What is the oldest date in the to_date column?
ANSWER:1985-03-01

In [284]:
from datetime import date
today = date.today()

In [285]:
# sort by to_date ascending
titles_df.sort_values(
    'to_date', ascending=True
).head(1)

emp_no     title   from_date     to_date
16064   20869  Engineer  1985-02-17  1985-03-01

### 11. What is the most recent date in the to_date column?
ANSWER: 2002-08-01



In [286]:
## mask of to_dates less than today:
# titles_df['to_date'] < today

# sort by and apply mask
titles_df.sort_values(
    'to_date', ascending=False
)[titles_df['to_date'] < today].head(1)

/var/folders/kj/13f127m163l12181fcrgh1vm0000gn/T/ipykernel_59029/4262389933.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  titles_df.sort_values(


emp_no     title   from_date     to_date
197364  233339  Engineer  1996-08-01  2002-08-01

## Exercises II

### 1. Copy the users and roles DataFrames from the examples above.



In [287]:
# Create users DataFrame
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users
# Create roles DataFrame
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})

### 2. What is the result of using a right join on the DataFrames?
ANSWER: A table is produced using information that was either in both DataFrames or the right DataFrame only. Left only data was dropped.

In [288]:
# Make right join:
joined_right = users.merge(roles, 
                           left_on='role_id', 
                           right_on='id', 
                           how='right', 
                           suffixes=['_from_users', '_from_roles'],
                           indicator=True)
joined_right

id_from_users name_from_users  role_id  id_from_roles name_from_roles  \
0            1.0             bob      1.0              1           admin   
1            2.0             joe      2.0              2          author   
2            3.0           sally      3.0              3        reviewer   
3            4.0            adam      3.0              3        reviewer   
4            NaN             NaN      NaN              4       commenter   

       _merge  
0        both  
1        both  
2        both  
3        both  
4  right_only

### 3. What is the result of using an outer join on the DataFrames?
ANSWER: All rows and columns from both DataFrames was included in the new DataFrame

In [289]:
# Make outer join:
joined_outer = users.merge(roles, 
                           left_on='role_id', 
                           right_on='id', 
                           how='outer', 
                           indicator=True)
joined_outer

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?
ANSWER: Only the 'id' and 'name' columns were outputted. They were empty of values.

In [290]:
# Drop foreign key (role_id) from users
users_dropped = users.drop(columns='role_id')

#Drop foreign key (id) from roles
roles_dropped = roles.drop(columns='id')

# Merge
dropped_fkey_merge = users_dropped.merge(roles_dropped)
dropped_fkey_merge

Empty DataFrame
Columns: [id, name]
Index: []

### 5. Load the mpg dataset from PyDataset.



In [291]:
# import data from pydataset and save 'mpg' as variable
from pydataset import data
mpg = data('mpg')

### 6. Output and read the documentation for the mpg dataset.



In [292]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




### 7. How many rows and columns are in the dataset?
ANSWER: There are 234 rows and 11 columns. This can be found either in the documentation or by .shape


### 8. Check out your column names and perform any cleanup you may want on them.



In [293]:
# Check out column names
mpg.columns

# Rename columns to full word
mpg = mpg.rename(columns={
    'displ': 'displacement',
    'cyl': 'cylinders',
    'trans': 'transmission',
    'drv': 'drive',
    'cty': 'city',
    'hwy': 'highway',
    'class': 'classification'
})

### 9. Display the summary statistics for the dataset.



In [294]:
# print .info()
print(mpg.info())
# print .describe()
print(mpg.describe())
# print .describe() for dtypes 'object'
print(mpg.loc[:, mpg.dtypes == 'object'].describe())

<class 'pandas.core.frame.DataFrame'>
Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   manufacturer    234 non-null    object 
 1   model           234 non-null    object 
 2   displacement    234 non-null    float64
 3   year            234 non-null    int64  
 4   cylinders       234 non-null    int64  
 5   transmission    234 non-null    object 
 6   drive           234 non-null    object 
 7   city            234 non-null    int64  
 8   highway         234 non-null    int64  
 9   fl              234 non-null    object 
 10  classification  234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB
None
       displacement         year   cylinders        city     highway
count    234.000000   234.000000  234.000000  234.000000  234.000000
mean       3.471795  2003.500000    5.888889   16.858974   23.440171
std        1.291959     4.509646    1.61

### 10. How many different manufacturers are there?
ANSWER: 15

In [295]:
mpg['manufacturer'].nunique()

15

### 11. How many different models are there?
ANSWER: 38

In [296]:
mpg['model'].nunique()

38

### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.



In [297]:
mpg['mileage_difference'] = (mpg['highway'] - mpg['city'])
mpg

manufacturer   model  displacement  year  cylinders transmission drive  \
1           audi      a4           1.8  1999          4     auto(l5)     f   
2           audi      a4           1.8  1999          4   manual(m5)     f   
3           audi      a4           2.0  2008          4   manual(m6)     f   
4           audi      a4           2.0  2008          4     auto(av)     f   
5           audi      a4           2.8  1999          6     auto(l5)     f   
..           ...     ...           ...   ...        ...          ...   ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)     f   
231   volkswagen  passat           2.0  2008          4   manual(m6)     f   
232   volkswagen  passat           2.8  1999          6     auto(l5)     f   
233   volkswagen  passat           2.8  1999          6   manual(m5)     f   
234   volkswagen  passat           3.6  2008          6     auto(s6)     f   

     city  highway fl classification  mileage_difference  
1      18       29  p        compact                  11  
2      21       29  p        compact                   8  
3      20       31  p        compact                  11  
4      21       30  p        compact                   9  
5      16       26  p        compact                  10  
..    ...      ... ..            ...                 ...  
230    19       28  p        midsize                   9  
231    21       29  p        midsize                   8  
232    16       26  p        midsize                  10  
233    18       26  p        midsize                   8  
234    17       26  p        midsize                   9  

[234 rows x 12 columns]

### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.



In [298]:
mpg['average_mileage'] = (mpg['highway'] + mpg['city'])/2
mpg

manufacturer   model  displacement  year  cylinders transmission drive  \
1           audi      a4           1.8  1999          4     auto(l5)     f   
2           audi      a4           1.8  1999          4   manual(m5)     f   
3           audi      a4           2.0  2008          4   manual(m6)     f   
4           audi      a4           2.0  2008          4     auto(av)     f   
5           audi      a4           2.8  1999          6     auto(l5)     f   
..           ...     ...           ...   ...        ...          ...   ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)     f   
231   volkswagen  passat           2.0  2008          4   manual(m6)     f   
232   volkswagen  passat           2.8  1999          6     auto(l5)     f   
233   volkswagen  passat           2.8  1999          6   manual(m5)     f   
234   volkswagen  passat           3.6  2008          6     auto(s6)     f   

     city  highway fl classification  mileage_difference  average_mileage  
1      18       29  p        compact                  11             23.5  
2      21       29  p        compact                   8             25.0  
3      20       31  p        compact                  11             25.5  
4      21       30  p        compact                   9             25.5  
5      16       26  p        compact                  10             21.0  
..    ...      ... ..            ...                 ...              ...  
230    19       28  p        midsize                   9             23.5  
231    21       29  p        midsize                   8             25.0  
232    16       26  p        midsize                  10             21.0  
233    18       26  p        midsize                   8             22.0  
234    17       26  p        midsize                   9             21.5  

[234 rows x 13 columns]

### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.



In [299]:
# View transmission types
mpg['transmission'].value_counts()

transmission
auto(l4)      83
manual(m5)    58
auto(l5)      39
manual(m6)    19
auto(s6)      16
auto(l6)       6
auto(av)       5
auto(s5)       3
auto(s4)       3
auto(l3)       2
Name: count, dtype: int64

In [300]:
# string method .contains
mpg['is_automatic'] = mpg.transmission.str.contains('auto')
mpg


manufacturer   model  displacement  year  cylinders transmission drive  \
1           audi      a4           1.8  1999          4     auto(l5)     f   
2           audi      a4           1.8  1999          4   manual(m5)     f   
3           audi      a4           2.0  2008          4   manual(m6)     f   
4           audi      a4           2.0  2008          4     auto(av)     f   
5           audi      a4           2.8  1999          6     auto(l5)     f   
..           ...     ...           ...   ...        ...          ...   ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)     f   
231   volkswagen  passat           2.0  2008          4   manual(m6)     f   
232   volkswagen  passat           2.8  1999          6     auto(l5)     f   
233   volkswagen  passat           2.8  1999          6   manual(m5)     f   
234   volkswagen  passat           3.6  2008          6     auto(s6)     f   

     city  highway fl classification  mileage_difference  average_mileage  \
1      18       29  p        compact                  11             23.5   
2      21       29  p        compact                   8             25.0   
3      20       31  p        compact                  11             25.5   
4      21       30  p        compact                   9             25.5   
5      16       26  p        compact                  10             21.0   
..    ...      ... ..            ...                 ...              ...   
230    19       28  p        midsize                   9             23.5   
231    21       29  p        midsize                   8             25.0   
232    16       26  p        midsize                  10             21.0   
233    18       26  p        midsize                   8             22.0   
234    17       26  p        midsize                   9             21.5   

     is_automatic  
1            True  
2           False  
3           False  
4            True  
5            True  
..            ...  
230          True  
231         False  
232          True  
233         False  
234          True  

[234 rows x 14 columns]

### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?
ANSWER: Honda, 28.5 mpg

In [301]:
mpg.groupby('manufacturer').average_mileage.agg('mean').sort_values(ascending=False).head(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

### 16. Do automatic or manual cars have better miles per gallon?
ANSWER: Manual Cars have better mpg

In [302]:
mpg.groupby('is_automatic').average_mileage.agg('mean')

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

## Exercises III

### 1. Use your get_db_url function to help you explore the data from the chipotle database.



In [303]:
# Create connection string with get_db_url
my_url = env.get_db_url(
    env.user,
    env.host,
    env.password,
    'chipotle')
# Write sql query
query = 'select * from orders'
# Create DataFrame
chipotle = pd.read_sql(query, my_url)
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

### 2. What is the total price for each order?



In [304]:
# change item_price dtype to numerical
item_p = chipotle['item_price'].str.replace('$', '')
item_p = item_p.str.replace(',', '')
item_p= item_p.astype('float')

# Add column for item_subtotal by multiplying quantity to item_price
chipotle['item_subtotal'] = chipotle['quantity'] * item_p

# groupby 'order_id' and sum item_subtotal
chipotle.groupby('order_id').item_subtotal.agg('sum')

order_id
1       11.56
2       33.96
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_subtotal, Length: 1834, dtype: float64

### 3. What are the most popular 3 items?
ANSWER: (Chicken Bowl, Chicken Burrito, Chips and Guacamole)

In [354]:
chipotle.groupby('item_name')['quantity'].sum().sort_values(ascending=False).head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

### 4. Which item has produced the most revenue?
ANSWER: Chicken Bowl

In [355]:
chipotle.groupby('item_name').item_subtotal.agg('sum').sort_values(ascending=False).head(1)

item_name
Chicken Bowl    8044.63
Name: item_subtotal, dtype: float64

### 5. Join the employees and titles DataFrames together.



In [356]:
# using .merge to join DataFrames together on emp_no
employee_titles = employees_df.merge(titles_df, left_on='emp_no', right_on='emp_no', how='inner', indicator=True)
employee_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date _merge  
0       Senior Engineer  1986-06-26  9999-01-01   both  
1                 Staff  1996-08-03  9999-01-01   both  
2       Senior Engineer  1995-12-03  9999-01-01   both  
3              Engineer  1986-12-01  1995-12-01   both  
4       Senior Engineer  1995-12-01  9999-01-01   both  
...                 ...         ...         ...    ...  
443303         Engineer  1987-08-30  1992-08-29   both  
443304  Senior Engineer  1992-08-29  9999-01-01   both  
443305     Senior Staff  1998-12-27  9999-01-01   both  
443306            Staff  1993-12-27  1998-12-27   both  
443307         Engineer  1997-11-30  9999-01-01   both  

[443308 rows x 10 columns]

### 6. For each title, find the hire date of the employee that was hired most recently with that title.


In [360]:
employee_titles.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)



In [318]:
# Because the titles df and departments df don't share a key I need to use a joiner table. I will use the dept_emp table.
# Write queries
query = '''
select *
from
titles'''
query2 = '''
select *
from
departments'''
query3 = '''
select emp_no, dept_no
from
dept_emp'''

# Create URL
my_url = env.get_db_url(
    env.user,
    env.host,
    env.password,
    'employees')

# Create DataFrames
df1 = pd.read_sql(query, my_url)
df2 = pd.read_sql(query2, my_url)
df3 = pd.read_sql(query3, my_url)

In [331]:
# Use .merge to join df1, df2, and df3
df = df1.merge(df3, left_on='emp_no', right_on='emp_no', how='inner', indicator=True)
df = df.merge(df2, left_on='dept_no', right_on='dept_no', how='inner')

In [348]:
# Use pd.crosstab to show the number of each title by department
pd.crosstab(df.dept_name, df.title, margins=True)

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                   298      2362        4             2027   
Development                       7769     58135        2            49326   
Finance                              0         0        2                0   
Human Resources                      0         0        2                0   
Marketing                            0         0        2                0   
Production                        6445     49649        4            42205   
Quality Management                1831     13852        4            11864   
Research                           378      2986        2             2570   
Sales                                0         0        2                0   
All                              16721    126984       24           107992   

title               Senior Staff   Staff  Technique Leader     All  
dept_name                                                           
Customer Service           13925   16150               309   35075  
Development                 1247    1424              7683  125586  
Finance                    12139   13929                 0   26070  
Human Resources            12274   14342                 0   26618  
Marketing                  13940   16196                 0   30138  
Production                  1270    1478              6557  107608  
Quality Management             0       0              1795   29346  
Research                   11637   13495               393   31461  
Sales                      36191   41808                 0   78001  
All                       102623  118822             16737  489903